pilot/pkg/serviceregistry/aggregate/controller.go

```diff
@@ -57,6 +57,55 @@ func (c *Controller) AddRegistry(registry Registry) {
 	c.registries = append(c.registries, registry)
 }
 
+// DeleteRegistry deletes specified registry from the aggregated controller
+func (c *Controller) DeleteRegistry(registry Registry) {
+	c.storeLock.Lock()
+	defer c.storeLock.Unlock()
+	if len(c.registries) == 0 {
+		log.Warnf("Registry list is empty, nothing to delete")
+		return
+	}
+	index, ok := c.GetRegistryIndex(registry)
+	if !ok {
+		log.Warnf("Registry is not found in the registries list, nothing to delete")
+		return
+	}
+	c.registries = append(c.registries[:index], c.registries[index+1:]...)
+}
+
+// RegistryElement is used to send Registry over a channel
+type RegistryElement struct {
+	index int
+	Registry
+}
+
+// GetRegistry returns Registry element in a protected maner
+func (c *Controller) GetRegistry() <-chan RegistryElement {
```

Comment:
GetRegistry is a very frequent operation, I don't think the overhead or complexity is acceptable.

All  that is needed for multicluster is an atomic way to get the pointer to the []registry - let's not make things more complicated for a corner case.

Reply:
I mean there are quite a bit of runtime costs to use a channel - and this is in the end the same (or worse) with the original patch that was holding a lock: you can't have 2 calls scanning at the same time, and the lock will be held until all are consumed.

So in effect is a very obfuscated and expensive way to sync all calls - which we didn't want in the first place.